module/zfs/zvol.c

```diff
@@ -966,13 +968,22 @@ zvol_request(struct request_queue *q, struct bio *bio)
 		 */
 		zvr->rl = zfs_range_lock(&zv->zv_range_lock, offset, size,
 		    RL_WRITER);
+		/*
+		 * FIXME: if there is at least another lock held on the same
+		 * block process this request synchronously.
+		 */
+		if (zfs_range_locked(&zv->zv_range_lock, P2ALIGN(offset,
+		    zv->zv_volblocksize), zv->zv_volblocksize) > 1)
+			need_sync = B_TRUE;
 		if (bio_is_discard(bio) || bio_is_secure_erase(bio)) {
-			if (zvol_request_sync || taskq_dispatch(zvol_taskq,
-			    zvol_discard, zvr, TQ_SLEEP) == TASKQID_INVALID)
+			if (zvol_request_sync || need_sync ||
+			    taskq_dispatch(zvol_taskq, zvol_discard, zvr,
+			    TQ_SLEEP) == TASKQID_INVALID)
```

Comment:
This is a different bug, but when inspecting `zvol_discard()` I see it's missing it's call to `zil_commit()`.  This is likely because it was accidentally dropped when porting this code from illumos since their block layer is structured differently.  We'll want to add it.